lib/charms/alertmanager_k8s/v0/alertmanager.py

```diff
@@ -0,0 +1,146 @@
+        """Returns a list of ip addresses of all the alertmanager units"""
+        return sorted(list(self._stored.alertmanagers.values()))
+
+    def _on_relation_departed(self, event: ops.charm.RelationDepartedEvent):
+        """This hook removes the address of the departing alertmanager from its local store.
+        This is needed for consumers such as prometheus, which should be aware of all alertmanager
+        instances.
+        """
+        if self._stored.alertmanagers.pop(event.unit.name, None):
+            # inform consumer about the change
+            # manually emitting the "available" event until scale-up/down "ready" flag is fixed
+            self.on.available.emit()
+
+    def _on_relation_broken(self, event: ops.charm.RelationBrokenEvent):
+        self._stored.alertmanagers.clear()
+        # inform consumer about the change
+        # manually emitting the "available" event until scale-up/down "ready" flag is fixed
+        self.on.available.emit()
+
+
+class AlertmanagerProvider(ProviderBase):
+    """A "provider" handler to be used by the Alertmanager charm for abstracting away all the
+    communication with consumers.
+    This provider auto-registers relation events on behalf of the main Alertmanager charm.
+
+    Arguments:
+            charm (CharmBase): consumer charm
+            service_name (str): a name for the provided service
+            consumes (dict): provider specifications
+            multi (bool): multiple relations flag
+
+    Attributes:
+            charm (CharmBase): the Alertmanager charm
+    """
+
+    _provider_relation_name = "alerting"
+    _public_api_port: int = 9093  # public port may not be the same as AlertmanagerCharm._api_port
+
+    def __init__(self, charm, service_name: str, version: str = None):
+        super().__init__(charm, self._provider_relation_name, service_name, version)
+        self.charm = charm
+        self._service_name = service_name
+
+        events = self.charm.on[self._provider_relation_name]
+        self.framework.observe(events.relation_joined, self._on_relation_joined)
+
+        # No need to observe `relation_departed` or `relation_broken`: data bags are auto-updated
+        # so both events are address on the consumer side.
+        self.framework.observe(events.relation_joined, self._on_relation_joined)
+
+    def _on_relation_joined(self, event: ops.charm.RelationJoinedEvent):
+        """This hook stores the public address of the newly-joined "alerting" relation in the
+        corresponding data bag.
+        This is needed for consumers such as prometheus, which should be aware of all alertmanager
+        instances.
+        """
+        # "ingress-address" is auto-populated incorrectly so rolling my own, "public_address"
+        event.relation.data[self.charm.unit]["public_address"] = "{}:{}".format(
+            self.model.get_binding(event.relation).network.bind_address, self._public_api_port
+        )
```

Comment:
Rather than pulling _public_api_port from the charm, the library should implement a "set_public_address" method that the charm calls on a relation_joined event. This is to make the library more portable is someone else wants to implement an alertmanager charm. They can reuse the library without needing to have specific private members of their class.

Reply:
Good point! Added.